python/sunbeam/deck/deck.py

```diff
@@ -0,0 +1,283 @@
+                assert isinstance(record, list), "All keyword records must be lists. Got " + str(type(record))
+                for item in record:
+                    assert isinstance(item, (int, float, str, _repeat)), " Keyword items must be either int, float, str, default or repeat. Got " + str(type(item))
+
+    def index( self, key ):
+        return next( (i for i, v in enumerate(self._kws) if v[0] == key) )
+
+    def indices( self, key ):
+        return [ i for i, v in enumerate(self._kws) if v[0] == key ]
+
+    def insert( self, key, v ):
+        if isinstance( key, int ) or isinstance( key, slice ):
+            self._kws.insert( key, v )
+        else:
+            raise TypeError( "Deck indices must be integers or slices, not " + type(key).__name__ )
+
+    @property
+    def select( self ):
+        return self._cursor_factory
+
+
+
+    def __repr__(self):
+        self.data_invariant()
+        string = ''
+        for kw in self._kws:
+            string += Deck.keyword_to_string( kw )
+            string += '\n'
+        return string
+
+    @staticmethod
+    def keyword_to_string( kw ):
+        string = ''
+        string += kw[0] + '\n'
+        for r in kw[1:]:
+            string += Deck.record_to_string(r)
+        return string
+
+    @staticmethod
+    def record_to_string( record ):
+        if (record == []):
+            return '/\n'
+        string = ''
+        line = ''
+        for item in record + ['/']:
+            if line and len(line + repr(item)) > 78:
+                string += '  ' + line + '\n'
+                line = ''
+            item_repr = repr(item) if item != '/' else '/'
+            line += item_repr + ' '
+        if line:
+            string += '  ' + line + '\n'
+        return string
+
+
+class _repeat:
+    def __init__(self, default=False, count=None, value=None):
+        self.default = default
+        self.count = count
+        self.value = value
+
+    def __rmul__(self, count):
+        r = _repeat(default=self.default, count=count, value=self.value)
+        r.data_invariant()
+        return r
+
+    def __repr__(self):
+        self.data_invariant()
+        return '{}*{}'.format(self.count if self.count else '',
+                              repr(self.value) if not self.default else '')
+
+    def data_invariant(self):
+        if not isinstance(self.default, bool):
+            raise ValueError("default parameter must be bool")
+        if self.count and not isinstance(self.count, int):
+            raise ValueError("Repeat count must be set and int")
+        if not self.default and not self.count:
+            raise ValueError("Cannot repeat non-default value without count")
+        if self.value and not isinstance(self.value, (int, float, str)):
+            raise ValueError("Values must be either int, float or str")
+        if not self.default and not self.value:
+            raise ValueError("Cannot repeat non default item with None value")
```

Comment:
Perhaps these should be either RuntimeError or asserts, but I don't have a strong opinion, so just leave it as-is.